torch_geometric/nn/models/pg_explainer.py

```diff
@@ -0,0 +1,270 @@
+        assert x.shape[0] == z.shape[0]
+        assert ~(batch is None and node_idxs is None)
+
+        self.model.eval()
+        clear_masks(self.model)
+        self.to(x.device)
+        optimizer = torch.optim.Adam(self.explainer_model.parameters(),
+                                     lr=self.lr)
+        self.explainer_model.train()
+
+        if self.log:  # pragma: no cover
+            pbar = tqdm(total=self.epochs)
+            pbar.set_description('Training Explainer')
+
+        bias = self.coeffs['bias']
+        if self.task == "graph":
+            # Get initial prediction.
+            prediction = self.get_initial_prediction(x, edge_index, batch,
+                                                     **kwargs)
+
+            assert (batch is not None) and x.shape[0] == batch.shape[0]
+            assert batch.unique().shape[0] == prediction.shape[0]
+            batch = batch.squeeze()
+            explainer_in = self._create_explainer_input(edge_index, z).detach()
+
+            for e in range(0, self.epochs):
+                optimizer.zero_grad()
+                t = self._get_temp(e)
+                self.edge_mask = self._compute_edge_mask(
+                    self.explainer_model(explainer_in), t, bias=bias)
+                set_masks(self.model, self.edge_mask, edge_index)
+                out = self.model(x=x, edge_index=edge_index, batch=batch,
+                                 **kwargs)
+                self.get_loss(out, prediction.squeeze(), node_idx=None,
+                              batch=batch, edge_index=edge_index).backward()
+                optimizer.step()
+                if self.log:  # pragma: no cover
+                    pbar.update(1)
+
+        else:
+            assert node_idxs.unique().shape[0] == node_idxs.shape[0]
+            for e in range(0, self.epochs):
+                loss = torch.tensor([0.0], device=x.device).detach()
+                t = self._get_temp(e)
+                optimizer.zero_grad()
+
+                for n in node_idxs:
+                    n = int(n)
+                    kwargs['z'] = z
+                    (x_n, edge_index_n, mapping, _, _,
+                     kwargs_n) = self.subgraph(n, x, edge_index, **kwargs)
+                    z_n = kwargs_n.pop('z')
+                    prediction = self.get_initial_prediction(
+                        x_n, edge_index_n, **kwargs_n)
+
+                    explainer_in = self._create_explainer_input(
+                        edge_index_n, z_n, mapping).detach()
+                    self.edge_mask = self._compute_edge_mask(
+                        self.explainer_model(explainer_in), t, bias=bias)
+                    # Should this be edge_index_n or edge_index?
+                    set_masks(self.model, self.edge_mask, edge_index_n)
+                    out = self.model(x=x_n, edge_index=edge_index_n,
+                                     **kwargs_n)
+                    loss += self.get_loss(out, prediction, mapping)
```

Comment:
One of my key concerns while implementing this was, the loss for the node classification task  is a sum across nodes. I am not sure how stable this would be if a large number of nodes are used to train `explainer_model`. @RBendias  any thoughts on this from your experience with explainers.

Reply:
This is the way they do it in the original [code](https://github.com/flyingdoog/PGExplainer/blob/ec57ea2db3b196da7eb93179acffc64ec6958828/BA-shapes.ipynb), right!? In the paper they use a rather small number of training examples (e.g. 5-30), maybe we could add that as a comment to the `node_idxs` argument. Does this concern the graph classification as well, since we use the sum in the loss function?

Reply:
Mentioning this in the comments is a good idea. And yeah this applies to both tasks.

Reply:
Sounds good! The explanation AUC for mutagenicity remains the same when taking the mean instead of the sum. Also, I tried some data sets where 100 training samples worked up to 4x better than 30. So maybe your first guess was correct, and we should switch e.g. to mean.